### ISSUE TITLE
Ability to bulk-delete old incomplete submissions 

### ISSUE THREAD

**Issue by pmangahis** [pkp/pkp-lib#6528]
**Describe the problem you would like to solve**
We've received 2 requests from hosted clients to have a multi-select option to remove incomplete submissions. 

**Describe the solution you'd like**
Currently, JMs need to delete incomplete submissions from authors and QuickSubmit plugin.

**Who is asking for this feature?**
Journal Editors, Journal Manager

**Comment by NateWr** [comment]
Thanks @pmangahis. Can you say a little bit more about the use-case and the steps that editors are taking? I assume that editors/managers are identifying old incomplete submissions in some way before deciding what to remove. How do they determine which ones to remove? Why do they feel they need to remove them?

**Comment by pmangahis** [comment]
In one case the journal was doing a manual import from BePress using the QuickSubmit plugin. I think when there is an error during the submission process using QuickSubmit it still creates a record. They want to be able to remove duplicates.

In the other journal, it seems that the journal has multiple pages (over 10) of incomplete submissions from what I can tell were the authors created that the JM would like to remove.

**Comment by NateWr** [comment]
Thanks! I think the QuickSubmit issue is probably something that we should address at its source. With the incomplete submissions, do you know if these were done by a single malicious user or spam bot? Or had they just accumulated over many years?

**Comment by pmangahis** [comment]
I think it might be both cases for the other journal.

**Comment by carolinatanigushi** [comment]
We also have this issue on our installations. 

Journals have previously let us know that there are a lot of abandoned submissions that they need to manually delete. It's an issue on our OPS installation as well, considering that out of our 357 currently active submissions, 239 have not been touched on the last six months.

**Comment by jnugent** [comment]
Just to add 0.02 to this from the systems side.  We've been able to  clean out incomplete submissions by using the following logic:

1.  We identify old and stale incomplete submissions by looking at the submissions table and querying based on submission_status > 0 and a date_last_activity that is older than a determined threshold with the client. 
2. These submission ids are put into a text file and iterated over on the command line in a shell script and passed to the tools/deleteSubmissions.php tool.
3. In 3.3, before we can do this we need to comment out the code in the submission file service that logs file deletions because of this issue https://github.com/pkp/pkp-lib/issues/7597

But this does seem to work for us.  We've a few clients that are very old and have a long history with OJS 2 and now have upwards of a thousand incomplete submissions accumulated over the years, so it's nice to be able to clean these out.

Jason

**Comment by librariam** [comment]
In addition to Jason's script suggestion above, for a use case with a smaller number of incomplete submissions - I wonder if it would make sense to bring back the old submission "Delete" button for admin accounts only (before a submission is declined)? I understand the reason for removing it for JE/JM accounts, but having an admin delete submissions quickly would be very helpful.

**Comment by Devika008** [comment]
Hello,

Here are the designs for the the issue! The considerations taken for the designs were:

- Adding a checkbox next to each incomplete submission in the submission list.
- Adding a checkbox to select all checkboxes from #1.
- Adding a delete button to delete all selected items.
- Adding a confirmation modal

Also I do like @librariam 's suggestion of only including this option for the admin only

**3.4**
![image](https://github.com/user-attachments/assets/f0705ba7-c23a-4170-81c9-0b5eee13f433)

On selecting a checkbox, the delete button becomes enabled
![image](https://github.com/user-attachments/assets/06b8308b-7bb4-4dd6-8438-b0192c2cf2b9)

You can select and deselect all after a checkbox has been selected
![image](https://github.com/user-attachments/assets/392b913d-bd8f-4ad0-9e39-546afc40a46d)

This needs to be an entirely intentional action and hence an extra layer of confirmation is necessary
![image](https://github.com/user-attachments/assets/c2d721ab-9dd2-4509-b3da-aa955f4e8fec)



**3.5**
For the 3.5 submission table, I confirmed with Israel that placing the button before, like in Gmail, is more intuitive. To make the table more accessible, the column heading can be visually hidden but still made available to screen reader users. When using a keyboard, there will be an "on focus" state. Developers can add an aria-label or a <span> with a class to hide it. The text to be announced by the element should be "Select/deselect the submission." I think the suggestion fits extremely well with the current proposed submission lists dashboard

![image](https://github.com/user-attachments/assets/fcf699ed-3e62-4c0a-8658-dbdb5f664f93)

![image](https://github.com/user-attachments/assets/b6d2a77f-d2c3-42f8-b382-4b07ac10def1)

![image](https://github.com/user-attachments/assets/b82607c0-c6cf-4ff9-af43-f4b52e3eedc0)

A confirmation message similar to 3.4 needs to be shown here as well. 

![image](https://github.com/user-attachments/assets/4784cf98-fd57-4cb0-9e98-12561e5ba8fc)

**Comment by taslangraham** [comment]
Thanks @Devika008

**Comment by taslangraham** [comment]
Ready for review.

### PRs

**pkp-lib (main)**
This introduces a new endpoint that accepts a list of submission IDs via a query param, performs the necessary validations, and deletes the corresponding submissions.
- pr: https://github.com/pkp/pkp-lib/pull/10258. @asmecher @eirikhanssen @jonasraoni

**UI library (main)** 
Please see demo [here](https://github.com/pkp/ui-library/pull/390#issue-2435859470)
- pr: https://github.com/pkp/ui-library/pull/390. @jardakotesovec

**submodule updates**
- ojs: https://github.com/pkp/ojs/pull/4387
- omp: https://github.com/pkp/omp/pull/1665
- ops: https://github.com/pkp/ops/pull/744

**Comment by jonasraoni** [comment]
@taslangraham I think you're tagging the wrong Erik, it should be @ewhanson 😁

**Comment by taslangraham** [comment]
> @taslangraham I think you're tagging the wrong Erik, it should be @ewhanson 😁

ha! you're right. thanks

**Comment by jardakotesovec** [comment]
@taslangraham Hi Taslan,
this unfortunately needs to go on hold for some time, at least the UI part.

Thing is that for version 3.5 (main branch) - the submission list panel is being replaced with new UI, which is not using ListPanel anymore (https://github.com/pkp/pkp-lib/issues/7495). At this addition can be considered once the new UI is more settled.  Therefore the changes you introduced to submission list panel would work only for version 3.4 - but at this point we are not adding new functionality to 3.4 as we need to focus on finishing things off for 3.5.

Sorry about that, this is a mistake on our part.

**Comment by taslangraham** [comment]
@jardakotesovec I understand. I'm fine picking this back up when the new UI is ready

**Comment by ewhanson** [comment]
In that case I'll also hold off on reviewing things until it's ready to move forward. Feel free to ping me again once it's time, @taslangraham.

**Comment by jardakotesovec** [comment]
ui-library PR for new dashboard - https://github.com/pkp/ui-library/pull/454
pkp-lib with some additional locales on top of @taslangraham PR - https://github.com/pkp/pkp-lib/pull/10655

@Devika008 Can you check it out on https://jarda.testing.publicknowledgeproject.org/ojs/ ? Currently it works for the admins. For authors it needs API adjustments.

@taslangraham Could you look into updating the endpoint so it would be also possible to delete incomplete submissions by their author?  Feel free to cherry-pick my commit with additional locales to your PR.  We can clean up the locale keys once we get feedback from Devika and Israel.

@israelcefrin Hi Israel, tagging you as well in case you have some suggestions on improving the accessibility of this feature. Please check out the notes below as well.

## Some notes:
### Logic
- Original intention was to allow bulk delete of incomplete submissions to admin, which would come from spammers. Given that we are not carrying over the 'Delete' button from 3.4 submission listing to the new one, we also decided to make it available for authors to delete theirs incomplete submissions as replacement.
- Additional use case which was covered by the Delete button in 3.4 submission listing, which allows to delete Declined submission to editors will be now awailable in workflow page - it will be added as part of existing [issue](https://github.com/pkp/pkp-lib/issues/10598?issue=pkp%7Cpkp-lib%7C10599).

### UI
- **Delete Incomplete Submission** button is present for admin on editorial dashboard and on author dashboard to be used by authors 
- **Delete Incomplete Submission** button is disabled if there is no incomplete submission that could be deleted in current listing
- **Delete Incomplete Submission** once there is incomplete submission that could be deleted, its using secondary styling - and once click on it its using 'active button' styling to indicate that this functionality is on - and the checkboxes shows up
- **Delete Incomplete Submission** change to warnable styling once some submission is selected and can be used to execute the deleation
- I used disabled checkboxes for submissions that can't be deleted, as I think this make it bit easier for SR to orient themself rather than empty cell
- From accessibility point of view - the tricky is that that we use one button first as switcher and than as normal button. Improvement probably should be that when it behaves as switcher - it should be coded as switcher as well.  Or maybe we could figure out to have separately switcher and the button to execute delelation - that would probably be improvement for SR.
- Column with checkboxes has SR label: **Select incomplete submissions to be deleted.**, hidden to normal users
- Checkboxes are connected with the submission title via aria-described-by - which seems to work quite nicely when testing with screen reader so its clear that its on the checkbox on which submission

**Comment by taslangraham** [comment]
Nice work @jardakotesovec! I'll take a look at updating the endpoint to support Authors today.

**Comment by israelcefrin** [comment]
Hello @jardakotesovec 

> From accessibility point of view - the tricky is that that we use one button first as switcher and than as normal button. Improvement probably should be that when it behaves as switcher - it should be coded as a switcher as well. Or maybe we could figure out to have separately switcher and the button to execute delelation - that would probably be an improvement for SR.

If you follow the first option you mentioned, then you will need to alert the users through some aria-live region alert to let them know that the element has changed its function. However, it might cause some confusion. I think that the second option, separating the switcher from the button, would be an easier solution and approach for those SR users. It would even ease the burden to people with cognitive disability because you wouldn't have the option to alert them that an element changed its function unless you trigger an alert (which would become annoying and intrusive each time the element changes).

> Checkboxes are connected with the submission title via aria-described-by - which seems to work quite nicely when testing with screen reader so its clear that its on the checkbox on which submission

We now have access to JAWS and NVDA  screen readers and can also test it. Could you please add me to the installation that is running this code? I can review it using these windows based screen readers along side VoiceOver.

**Comment by asmecher** [comment]
Related: https://github.com/pkp/pkp-lib/issues/4348

**Comment by jardakotesovec** [comment]
After discussion with Devika and Alec, here are some tweaks to work on:

1) Allowing Journal Manager to also bulk delete incomplete submissions, so they don't have to communicate submissions to be deleted always with admin.
@taslangraham Could you make this adjustment on the API please?

2) To address some of the UI concerns with double role of the Delete Incomplete Submissions button - [here](https://www.figma.com/design/Wf7sDlUg2372jaKKTJ0Mgz/OJS-3.4-3.5?node-id=10432-10972&t=aZEbmAIHbx1ouUSa-4) are the changes, which I will work on.

3) Will make the disabled checkboxes screen reader only, so its cleaner UI for users who are not using screen reader.

**Comment by taslangraham** [comment]
> 1. Allowing Journal Manager to also bulk delete incomplete submissions, so they don't have to communicate submissions to be deleted always with admin.
>    [@taslangraham](https://github.com/taslangraham) Could you make this adjustment on the API please?

I've updated the API to allow managers to delete incomplete submissions

**Comment by jardakotesovec** [comment]
@taslangraham Yeah.. I was also thinking similar thing when I was writing down the second option :-).   Will let @Devika008 decide whether it's worth adjusting.

**Comment by Devika008** [comment]
The second option is what we should go with! Taslan is right that the users would like to see consistency and not have to wonder why something is disappearing

**Comment by jardakotesovec** [comment]
Oki, updated.

ui-library pr: https://github.com/pkp/ui-library/pull/461

Locale keys that are used:
```
admin.submissions.incomplete.bulkDelete.confirm
admin.submissions.incomplete.bulkDelete.body
admin.submissions.incomplete.bulkDelete.column.description
admin.submissions.incomplete.bulkDelete.button
```

@taslangraham Please just check whether there are some additional locale keys that could be removed in pkp-lib and let me know once your PR is ready to be merged.

**Comment by taslangraham** [comment]
> 
> [@taslangraham](https://github.com/taslangraham) Please just check whether there are some additional locale keys that could be removed in pkp-lib and let me know once your PR is ready to be merged.


I've removed the unused locale keys, and moved the ones in use over to `submisisons.po` as discussed on Mattermost. 

@ewhanson do you mind giving the [PR](https://github.com/pkp/pkp-lib/pull/10258/files) a review before I merge? Thanks!

**Comment by ewhanson** [comment]
Thanks @taslangraham. Looks good to me!

**Comment by taslangraham** [comment]
Thanks @ewhanson.
@jardakotesovec I'll be merging the PKP lib PR.

**Comment by jardakotesovec** [comment]
@taslangraham Thanks, ui-library also merged.

@Tribunal33 Would you mind do some QAing on this one?

To summarise what we have intended to implement, feature is ability to delete incomplete submissions:
- as author I can delete my incomplete submissions 
- as admin or journal manager I can delete incomplete submissions in the journal I have such roles

**Comment by Tribunal33** [comment]
Taking a look now @jardakotesovec and @taslangraham  and will write up some test cases for it.

**Comment by Tribunal33** [comment]
This looks good.  @taslangraham can we add some scope creep?  I am not seeing a convenient way to cancel the process once the "Delete Incomplete Submissions" button has been engaged.  I have a solution.  Treat it like a filter with incomplete stage as the option and have the filter be deletable like other filters?  This would solve two problems.  First one is for those journals with pages and pages of submissions with intersparsed incompletes submissions it would bring them all to the top of the list.  Second, there is already a **clear filter** button that can be applied to remove the filter.  Here is a mockup of what I'm talking about.  Let me know if there are any questions. 

![Image](https://github.com/user-attachments/assets/35b8941a-1721-4fad-930e-895e0c4512c6)

**Comment by jardakotesovec** [comment]
@Tribunal33  Or 'Cancel' button can show up next to Delete Incomplete Submissions, which was my suggestion. But I think its minor concern as its infrequent action - thats probably why @Devika008 kept it simple. But I will let her respond to it.

**Comment by Tribunal33** [comment]
After talking with @Devika008 and @taslangraham we will go with an additional **Cancel** button beside the **Delete Incomplete Submissions** button to reset this action.  We will ignore the filtering idea as that breaks a PKP rule for filters.  I believe Devika has upgrade the Figma Files to reflect this change : 

![Image](https://github.com/user-attachments/assets/11610289-8847-4c68-aed6-25d24aeb7d55)

**Comment by jardakotesovec** [comment]
@Tribunal33 Oki, updated on main. Feel free to close this one if you don't see any other issue here.

**Comment by Tribunal33** [comment]
@taslangraham and @jardakotesovec I am sorry to do this to ya but there is one more issue.  If you have incomplete submissions on multiple pages that have been selected and click on the delete incomplete submissions button it only deletes from the page that you initialially selected.  To test this:
precondition :  you will need to create enough incomplete submissions to create another page
1. Then try initially starting on page 1 and selecting the three dots
2. Select Delete Incomplete submissions
3. Select an incomplete submission from page 1
4. Select **Next Page**
5. Select an incomplete submission from page 2
6. Select Delete Incomplete submissions button

**Expected outcome:** both incomplete submissions should have been removed from the filter list
**Actual outcome**: only page 1 (or initial pages) incomplete submissions will be removed.

**Comment by jardakotesovec** [comment]
Right.. from quick check the issue is not pagination. But the fact that our useFetch (which is using ofetch on background) is encoding array query param as `/_submissions?ids=20&ids=21`, which is not PHP backend friendly.. as it needs to be I think `/_submissions?ids[]=20&ids[]=21`. So we will need to adjust useFetch to encode arrays in php friendly way.. 

So I think the behaviour is that it just deletes the last item on that list..

**Comment by jardakotesovec** [comment]
Seems that they intend to address it in future versions by leveraging URLSearchParams

https://github.com/unjs/ufo/issues/185
https://github.com/unjs/ufo/issues/208
https://github.com/unjs/ofetch/pull/440

@taslangraham Could you please try to find some workaround until this is resolved upstream? Ideally to address it somewhere within useFetch.js composable.

**Comment by taslangraham** [comment]
@jardakotesovec I'll take a look

**Comment by taslangraham** [comment]
@jardakotesovec I added a PR with the workaround here - https://github.com/pkp/ui-library/pull/486. Please take a look

Approach I took was to rename the array query params to include `[]` before passing to `ofetchInstance`. So a query  entry like `{ids: [1,2,3]}` would become `{'ids[]':[1,2,3]}`

**Comment by taslangraham** [comment]
hmm, tests are failing. Looking into it

**Comment by taslangraham** [comment]
> hmm, tests are failing. Looking into it

fixed

**Comment by jardakotesovec** [comment]
@taslangraham That looks great, thank you! Feel free to merge it. And than we can ping @Tribunal33 if he can still find something :-).

**Comment by Tribunal33** [comment]
@taslangraham and @jardakotesovec this passes QA.  To test the remaining bug I needed to create enough incomplete submissions to see them on multiple pages in the submission list views. 
Here are the High level test ideas: 
1. Remove incomplete submission – single submission
2. Remove incomplete submission -- multi submission
3. Default button functionality
4. Button option visible from various roles
5. Remove incomplete submissions from different list pages
a. Each submission list has 30 submissions before the next page of submissions
